### PR TITLE
Option to duplicate profile with its permissions

### LINKED ISSUE #16290: Option to duplicate team profiles: suggestion
I would like to have the option to duplicate a team profile in back office. 

This would speed things up :)

### PR DESCRIPTION
**Pull request by kpodemski**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Thanks to this PR, you can easily duplicate employee profiles
| Type?             | new feature
| Category?         | BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #16290.
| How to test?      | Follow instructions from Issue
| Possible impacts? | Please indicate what parts of the software we need to check to make sure everything is alright.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27966)

### CODE REVIEW

**Review by PierreRambaud — COMMENTED** [review]
Can we wait until the PR about the migrated page is merged?

**Comment by kpodemski** [comment]
@PierreRambaud sure 👍

File: src/Adapter/Profile/CommandHandler/BulkDuplicateProfileHandler.php [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Profile\CommandHandler;
```
Comment (Progi1984) [inline src/Adapter/Profile/CommandHandler/BulkDuplicateProfileHandler.php:27]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Adapter\Profile\CommandHandler;
```

File: src/Adapter/Profile/CommandHandler/BulkDuplicateProfileHandler.php [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+use PrestaShop\PrestaShop\Core\Domain\Profile\Exception\ProfileNotFoundException;
+use PrestaShopException;
+use Profile;
+
+/**
+ * Class BulkDuplicateProfileHandler
+ *
+ * @internal
+ */
+final class BulkDuplicateProfileHandler extends AbstractProfileHandler implements BulkDuplicateProfileHandlerInterface
```
Comment (Progi1984) [inline src/Adapter/Profile/CommandHandler/BulkDuplicateProfileHandler.php:43]:
```suggestion
class BulkDuplicateProfileHandler extends AbstractProfileHandler implements BulkDuplicateProfileHandlerInterface
```

File: src/Adapter/Profile/CommandHandler/BulkDuplicateProfileHandler.php [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+use PrestaShop\PrestaShop\Core\Domain\Profile\Exception\FailedToDuplicateProfileException;
+use PrestaShop\PrestaShop\Core\Domain\Profile\Exception\ProfileException;
+use PrestaShop\PrestaShop\Core\Domain\Profile\Exception\ProfileNotFoundException;
+use PrestaShopException;
+use Profile;
+
+/**
+ * Class BulkDuplicateProfileHandler
+ *
+ * @internal
```
Comment (Progi1984) [inline src/Adapter/Profile/CommandHandler/BulkDuplicateProfileHandler.php:41]:
```suggestion
 * Class BulkDuplicateProfileHandler
```

File: src/Adapter/Profile/CommandHandler/DuplicateProfileHandler.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+use PrestaShop\PrestaShop\Core\Domain\Profile\Exception\ProfileNotFoundException;
+use PrestaShopException;
+use Profile;
+
+/**
+ * Class DuplicateProfileHandler
+ *
+ * @internal
+ */
+final class DuplicateProfileHandler extends AbstractProfileHandler implements DuplicateProfileHandlerInterface
```
Comment (Progi1984) [inline src/Adapter/Profile/CommandHandler/DuplicateProfileHandler.php:43]:
```suggestion
 */
class DuplicateProfileHandler extends AbstractProfileHandler implements DuplicateProfileHandlerInterface
```

File: src/Core/Domain/Profile/Command/BulkDuplicateProfileCommand.php [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+     */
+    public function __construct(array $profileIds)
+    {
+        $this->setProfileIds($profileIds);
+    }
+
+    /**
+     * @return ProfileId[]
+     */
+    public function getProfileIds()
```
Comment (Progi1984) [inline src/Core/Domain/Profile/Command/BulkDuplicateProfileCommand.php:52]:
```suggestion
    public function getProfileIds(): array
```

File: src/Core/Domain/Profile/Command/BulkDuplicateProfileCommand.php [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    {
+        return $this->profileIds;
+    }
+
+    /**
+     * @param array $profileIds
+     */
+    private function setProfileIds(array $profileIds)
```
Comment (Progi1984) [inline src/Core/Domain/Profile/Command/BulkDuplicateProfileCommand.php:60]:
```suggestion
    private function setProfileIds(array $profileIds): void
```

File: src/Core/Domain/Profile/Command/BulkDuplicateProfileCommand.php [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Profile\Command;
```
Comment (Progi1984) [inline src/Core/Domain/Profile/Command/BulkDuplicateProfileCommand.php:27]:
```suggestion
 */

declare(strict_types=1);

namespace PrestaShop\PrestaShop\Core\Domain\Profile\Command;
```

File: src/Core/Domain/Profile/Command/DuplicateProfileCommand.php [hunk 8]
```diff
@@ -0,0 +1,56 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Profile\Command;
```
Comment (Progi1984) [inline src/Core/Domain/Profile/Command/DuplicateProfileCommand.php:27]:
```suggestion
 */

declare(strict_types=1):

namespace PrestaShop\PrestaShop\Core\Domain\Profile\Command;
```

File: src/Core/Domain/Profile/Command/DuplicateProfileCommand.php [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+{
+    /**
+     * @var ProfileId
+     */
+    private $profileId;
+
+    /**
+     * @param int $profileId
+     */
+    public function __construct($profileId)
```
Comment (Progi1984) [inline src/Core/Domain/Profile/Command/DuplicateProfileCommand.php:44]:
```suggestion
    public function __construct(int $profileId)
```

File: src/Core/Domain/Profile/Command/DuplicateProfileCommand.php [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+    {
+        $this->profileId = new ProfileId((int) $profileId);
```
Comment (Progi1984) [inline src/Core/Domain/Profile/Command/DuplicateProfileCommand.php:46]:
```suggestion
        $this->profileId = new ProfileId($profileId);
```

File: src/Core/Domain/Profile/Command/DuplicateProfileCommand.php [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    }
+
+    /**
+     * @return ProfileId
+     */
+    public function getProfileId()
```
Comment (Progi1984) [inline src/Core/Domain/Profile/Command/DuplicateProfileCommand.php:52]:
```suggestion
    public function getProfileId(): ProfileId
```

File: src/Core/Domain/Profile/CommandHandler/BulkDuplicateProfileHandlerInterface.php [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+ * Interface BulkDuplicateProfileHandlerInterface defines profile bulk deletion handler.
+ */
+interface BulkDuplicateProfileHandlerInterface
+{
+    /**
+     * Duplicate multiple profiles.
+     *
+     * @param BulkDuplicateProfileCommand $command
+     */
+    public function handle(BulkDuplicateProfileCommand $command);
```
Comment (Progi1984) [inline src/Core/Domain/Profile/CommandHandler/BulkDuplicateProfileHandlerInterface.php:41]:
```suggestion
    public function handle(BulkDuplicateProfileCommand $command): void;
```

File: src/Core/Domain/Profile/CommandHandler/DuplicateProfileHandlerInterface.php [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ * Interface DuplicateProfileHandlerInterface defines profile deletion handler.
+ */
+interface DuplicateProfileHandlerInterface
+{
+    /**
+     * Duplicate profile.
+     *
+     * @param DuplicateProfileCommand $command
+     */
+    public function handle(DuplicateProfileCommand $command);
```
Comment (Progi1984) [inline src/Core/Domain/Profile/CommandHandler/DuplicateProfileHandlerInterface.php:41]:
```suggestion
    public function handle(DuplicateProfileCommand $command): void;
```

File: src/Core/Domain/Profile/Exception/FailedToDuplicateProfileException.php [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Domain\Profile\Exception;
```
Comment (Progi1984) [inline src/Core/Domain/Profile/Exception/FailedToDuplicateProfileException.php:27]:
```suggestion
 */

declare(strict_types=1)

namespace PrestaShop\PrestaShop\Core\Domain\Profile\Exception;
```

File: src/Core/Grid/Definition/Factory/BulkDuplicateActionTrait.php [hunk 15]
```diff
@@ -0,0 +1,80 @@
...
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
+
+namespace PrestaShop\PrestaShop\Core\Grid\Definition\Factory;
```
Comment (Progi1984) [inline src/Core/Grid/Definition/Factory/BulkDuplicateActionTrait.php:27]:
```suggestion
 */

declare(strict_types=1)

namespace PrestaShop\PrestaShop\Core\Grid\Definition\Factory;
```

File: src/Core/Grid/Definition/Factory/BulkDuplicateActionTrait.php [hunk 16]
```diff
@@ -0,0 +1,80 @@
...
+ */
+trait BulkDuplicateActionTrait
+{
+    /**
+     * @param string $bulkDuplicateRouteName
+     * @param array $options
+     *
+     * @return BulkActionInterface
+     */
+    protected function buildBulkDuplicateAction(string $bulkDuplicateRouteName, array $options = [])
```
Comment (Progi1984) [inline src/Core/Grid/Definition/Factory/BulkDuplicateActionTrait.php:44]:
```suggestion
    protected function buildBulkDuplicateAction(string $bulkDuplicateRouteName, array $options = []): SubmitBulkAction
```

File: src/Core/Grid/Definition/Factory/BulkDuplicateActionTrait.php [hunk 17]
```diff
@@ -0,0 +1,80 @@
...
+    /**
+     * Shortcut method to translate text.
+     *
+     * @param string $id
+     * @param array $options
+     * @param string $domain
+     *
+     * @return string
+     */
+    abstract protected function trans($id, array $options, $domain);
```
Comment (Progi1984) [inline src/Core/Grid/Definition/Factory/BulkDuplicateActionTrait.php:79]:
```suggestion
    abstract protected function trans(string $id, array $options, string $domain): string;
```

File: src/Core/Grid/Definition/Factory/DuplicateActionTrait.php [hunk 18]
```diff
@@ -0,0 +1,80 @@
...
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ */
```
Comment (Progi1984) [inline src/Core/Grid/Definition/Factory/DuplicateActionTrait.php:25]:
```suggestion
 */

```

File: src/Core/Grid/Definition/Factory/DuplicateActionTrait.php [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+            ->setName($this->trans('Duplicate', [], 'Admin.Actions'))
+            ->setIcon('content_copy')
+            ->setOptions($options)
+        ;
+    }
+
+    /**
+     * Shortcut method to translate text.
+     */
+    abstract protected function trans($id, array $options, $domain);
```
Comment (Progi1984) [inline src/Core/Grid/Definition/Factory/DuplicateActionTrait.php:79]:
```suggestion
    abstract protected function trans(string $id, array $options, string $domain): string;
```

File: src/Core/Grid/Definition/Factory/DuplicateActionTrait.php [hunk 20]
```diff
@@ -0,0 +1,80 @@
...
+trait DuplicateActionTrait
+{
+    protected function buildDuplicateAction(
+        string $duplicateRouteName,
+        string $duplicateRouteParamName,
+        string $duplicateRouteParamField,
+        string $method = 'POST',
+        array $extraRouteParams = [],
+        array $options = []
+    ): RowActionInterface {
```
Comment (Progi1984) [inline src/Core/Grid/Definition/Factory/DuplicateActionTrait.php:46]:
```suggestion
    ): SubmitRowAction {
```

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfileController.php [hunk 21]
```diff
@@ -248,6 +250,34 @@ public function deleteAction($profileId)
...
+     *     "is_granted('delete', request.get('_legacy_controller')~'_')",
+     *     message="You do not have permission to edit this."
+     * )
+     * @DemoRestricted(redirectRoute="admin_profiles_index")
+     *
+     * @param int $profileId
+     *
+     * @return RedirectResponse
+     */
+    public function duplicateAction($profileId)
```
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfileController.php:266]:
```suggestion
    public function duplicateAction(int $profileId): RedirectResponse
```

File: src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfileController.php [hunk 22]
```diff
@@ -278,6 +308,33 @@ public function bulkDeleteAction(Request $request)
...
+     * Bulk duplicate profiles.
+     *
+     * @AdminSecurity("is_granted('create', request.get('_legacy_controller'))")
+     * @DemoRestricted(redirectRoute="admin_profiles_index")
+     *
+     * @param Request $request
+     *
+     * @return RedirectResponse
+     */
+    public function bulkDuplicateAction(Request $request)
```
Comment (Progi1984) [inline src/PrestaShopBundle/Controller/Admin/Configure/AdvancedParameters/ProfileController.php:321]:
```suggestion
    public function bulkDuplicateAction(Request $request): RedirectResponse
```

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@kpodemski Could you add Behat tests, please ?

**Comment by Julievrz** [comment]
Wording suggestion:

**Modal**

Title: Duplicate selection?
Body: [No body]
Exit buttons: Cancel/Duplicate (Admin.Actions)

(Although I'm wondering if a modal is really necessary here. It's an additional step for the user when there's no big risk. It's not like he/she is about to delete a bunch of items for good.)

**Label**

About the label of the bulk action, I think the whole wording/design could be improved but in the meantime, I suggest keeping the existing string "Duplicate selection" in Admin.Actions, instead of creating the string "Duplicate selected". (It's also more consistent with other pages).

**Comment by kpodemski** [comment]
@PierreRambaud can I continue with this one? :>

**Review by Progi1984 — CHANGES_REQUESTED** [review]
@kpodemski The migrated PR has been merged so you can continue this one (and rebase it).

**Comment by matks** [comment]
@kpodemski do we keep this PR open or do we close? I'm a little sorry but it seems you could not find time to work on it

**Comment by kpodemski** [comment]
@matks 

keep it open, please ;-)

**Comment by kpodemski** [comment]
@Progi1984, would you like me to add a test if the permissions are the same between a source profile and a profile that has been cloned?

**Comment by Progi1984** [comment]
@kpodemski It should be perfect.

**Comment by matks** [comment]
Hi @kpodemski I believe this draft PR is now too stale. I close it for now, just to help navigating old PRs. But we can reopen it anytime you want 😉 obviously. A closed PR does not mean it has to stay closed forever 💪 ! I just do this to make it easier to distinguish old PRs that might be merged one day and PRs that - unfortunately - I think are too old and will never be merged.

Thank you anyway for submitting the PR 👍

**Pull request closed without merging**